compiler/resolution/nilChecking.cpp

```diff
@@ -0,0 +1,782 @@
+    if (storeToGlobals || storeToUnknown) {
+      for_vector(Symbol, sym, symbols) {
+        if (sym->isConstValWillNotChange() == false) {
+          if (storeToUnknown || isOuterVar(sym, fn))
+            OUT[sym] = unknownAliasLocation();
+        }
+      }
+    }
+
+    if (trace) {
+      printAliasMap("+-", OUT, traceCopy);
+      delete traceCopy;
+    }
+  }
+}
+
+void findNilDereferences(FnSymbol* fn) {
+
+  bool debugging = 0 == strcmp(fn->name, debugNilsForFn) ||
+                   fn->id == debugNilsForId;
+
+  if (debugging) {
+    printf("Visiting function %s id %i\n", fn->name, fn->id);
+    nprint_view(fn);
+    gdbShouldBreakHere();
+  }
+
+  BasicBlock::buildBasicBlocks(fn);
+
+  // Gather the variables to consider
+  std::vector<Symbol*> idxToSym;
+  std::map<Symbol*, int> symToIdx;
+
+  {
+    int index = 0;
+    std::vector<SymExpr*> symExprs;
+    collectSymExprs(fn, symExprs);
+    for_vector(SymExpr, se, symExprs) {
+      Symbol* sym = se->symbol();
+      if (isArgSymbol(sym) || isVarSymbol(sym)) {
+        if (symToIdx.count(sym) == 0) {
+          // Add it to the map/array
+          symToIdx.insert(std::pair<Symbol*, int>(sym, index));
+          idxToSym.push_back(sym);
+          INT_ASSERT((int)(idxToSym.size()-1) == index);
+          index++;
+          if (debugging) {
+            printf("Tracking symbol %i %s\n", sym->id, sym->name);
+          }
+        }
+      }
+    }
+  }
+
+  size_t nbbs = fn->basicBlocks->size();
+
+  // allocate the IN and OUT data flow sets
+  std::vector<AliasMap> IN(nbbs);
+  std::vector<AliasMap> OUT(nbbs);
+
+  bool changed = false;
+
+  do {
+
+    changed = false;
+
+    for (size_t i = 0; i < nbbs; i++) {
+      if (debugging) {
+        printf("running checkBasicBlock on bb %i\n", (int)i);
+      }
+      checkBasicBlock(fn, (*fn->basicBlocks)[i], idxToSym, IN[i], OUT[i],
+                      /*raise errors?*/ false);
+    }
+
+    // for each basic block, update IN block based on predecessors OUT blocks
+    for (size_t i = 0; i < nbbs; i++) {
+      BasicBlock* bb = (*fn->basicBlocks)[i];
+
+      AliasMap nextIn;
+
+      for_vector(BasicBlock, bbin, bb->ins) {
+        const AliasMap& from = OUT[bbin->id];
+        for (AliasMap::const_iterator it = from.begin();
+             it != from.end();
+             ++it) {
+          Symbol* sym = it->first;
+          AliasLocation loc = it->second;
+          if (nextIn.count(sym) == 0) {
+            nextIn[sym] = loc;
+          } else {
+            AliasLocation was = nextIn[sym];
+            // combine alias locations
+            if (was.type == loc.type &&
+                was.location == loc.location) {
+              // OK, do nothing
+            } else if (was.type == MUST_ALIAS_IGNORED) {
+              nextIn[sym] = loc;
+            } else {
+              // Conflicting input -> unknown
+              nextIn[sym] = unknownAliasLocation();
+            }
+          }
+        }
+      }
```

Comment:
Again, for readability, write the above as, ex. `AliasMap nextIn = combineIns(bb);`.